File: lib/generate.go
package lib

import (
	"familytree-gen/dev/assets"
	"familytree-gen/dev/types"
	"math/rand"
)

func GenerateAPerson(_country *string, _sex *bool) types.Person {
	var country string
	var sex bool

	if _country == nil {
		country = getRandomCountry()
	} else {
		country = *_country
	}

	if _sex == nil {
		sex = rand.Float32() < 0.5
	} else {
		sex = *_sex
	}

	forename := randomForename(country, sex)
	surname := randomSurname(country)
}

func randomSurname(country string) string {
	surnames := assets.SurnamesData[country][rand.Intn(len(assets.SurnamesData[country]))]

	return surnames.Localized[0]
}

func randomForename(country string, sex bool) string {
	names := assets.ForenamesData[country][rand.Intn(len(assets.ForenamesData[country]))].Names

	var filteredNames []types.Forenames

	for _, name := range names {
		condition := name.Gender == "M"
		if condition == sex {
			filteredNames = append(filteredNames, name)
		}
	}

	randName := filteredNames[rand.Intn(len(filteredNames))].Localized

	return randName[0]
}

func getRandomCountry() string {
	values := make([]string, 0, len(assets.CountriesCodeData))
	for _, v := range assets.CountriesCodeData {
		values = append(values, v)
	}

	return values[rand.Intn(len(values))]
}

File: src/lib/generate.ts
import { Codes } from "../assets/ISO3166-1.alpha2";
import Forenames from "../assets/common-forenames-by-country.min.json";
import Surenames from "../assets/common-surnames-by-country.min.json";

export function randomEnumKey<T extends Record<string, string | number>>(
	anEnum: T
): keyof T {
	const enumKeys = Object.keys(anEnum) as (keyof T)[];
	const randomIndex = Math.floor(Math.random() * enumKeys.length);
	const randomEnumKey = enumKeys[randomIndex];
	return randomEnumKey;
}

export function generateAFamily(country: string[], config?: {
	childrenThreshold?: number;
	maleThreshold?: number;
	bothParentsSameCountry?: boolean;
	mothersCountry?: string[];
    id?: number;
}) {
	const rndValue = Math.random() < 0.5;
    if (!country) country = [randomEnumKey(Codes)];
	if (!config) {
		config = {
			childrenThreshold: Math.random(),
			maleThreshold: Math.random(),
			bothParentsSameCountry: rndValue,
			mothersCountry: rndValue ? country : [Codes[randomEnumKey(Codes)]],
            id: 0,
		};
	} else {
		config = {
			childrenThreshold: Math.random(),
			maleThreshold: Math.random(),
			bothParentsSameCountry: rndValue,
			mothersCountry: rndValue ? country : [Codes[randomEnumKey(Codes)]],
            id: 0,
			...config,
		};
	}
}

function generateAPerson(country: Codes[], config?: {
    isParent: boolean
}) {

}

function generateAName(country: string[], sex: string) {
    let surname = "";
    let forename = "";
    const randomCountry: string = country[Math.round(Math.random() * country.length)] ?? country[0];
	const SurnameCountry = Surenames[randomCountry as keyof typeof Surenames];
	const ForenamesCountry = Forenames[randomCountry as keyof typeof Forenames];
    if(sex == "Female") {
		const name = SurnameCountry[Math.round(Math.random() * SurnameCountry.length)].localized;
    } else {

    }
}

function getEnumKey(country: Codes) {
    console.log(country);
}